packages/matter.js/src/protocol/interaction/InteractionServer.ts

```diff
@@ -371,17 +388,248 @@ export function attributePathToId({ endpointId, clusterId, attributeId }: TypeFr
+
+    public clear() {
+        this.endpoints.clear();
+        this.attributes.clear();
+        this.attributePaths.length = 0;
+        this.events.clear();
+        this.eventPaths.length = 0;
+        this.commands.clear();
+        this.commandPaths.length = 0;
+    }
+
+    public initializeFromEndpoint(endpoint: Endpoint) {
+        this.clear();
+
+        this.verifyAndInitializeStructureElementsFromEndpoint(endpoint); // Initialize Data from Root Endpoint
+        this.initializeStructureFromEndpoints(endpoint); // Initialize Data from Child Endpoints
+    }
+
+    private initializeStructureFromEndpoints(endpoint: Endpoint) {
+        const endpoints = endpoint.getChildEndpoints();
+        for (let endpointIndex = 0; endpointIndex < endpoints.length; endpointIndex++) {
+            this.verifyAndInitializeStructureElementsFromEndpoint(endpoints[endpointIndex]);
+            this.initializeStructureFromEndpoints(endpoints[endpointIndex]);
+        }
+    }
+
+    private verifyAndInitializeStructureElementsFromEndpoint(endpoint: Endpoint) {
+        if (endpoint.id === undefined) {
+            throw new Error(`Endpoint ID is undefined. It needs to be initialized first!`);
+        }
+
+        endpoint.verifyRequiredClusters();
+
+        for (const cluster of endpoint.getAllClusterServers()) {
+            const { id: clusterId, attributes: clusterAttributes, _events: clusterEvents, _commands: clusterCommands } = cluster;
+            // Add attributes
+            for (const name in clusterAttributes) {
+                const attribute = clusterAttributes[name];
+                const path = { endpointId: endpoint.id, clusterId, attributeId: attribute.id };
+                this.attributes.set(attributePathToId(path), attribute);
+                this.attributePaths.push(path);
+            }
+
+            // Add events
+            for (const name in clusterEvents) {
+                const event = clusterEvents[name];
+                const path = { endpointId: endpoint.id, clusterId, eventId: event.id };
+                this.events.set(eventPathToId(path), event);
+                this.eventPaths.push(path);
+            }
+
+            // Add commands
+            for (const name in clusterCommands) {
+                const command = clusterCommands[name];
+                const path = { endpointId: endpoint.id, clusterId, commandId: command.invokeId };
+                this.commands.set(commandPathToId(path), command);
+                this.commandPaths.push(path);
+            }
+        }
+
+        if (this.endpoints.has(endpoint.id)) throw new Error(`Endpoint ID ${endpoint.id} exists twice`);
+
+        this.endpoints.set(endpoint.id, endpoint);
+    }
+
+    toHex(value: number | undefined) {
+        return value === undefined ? "*" : `0x${value.toString(16)}`;
+    }
+
+    resolveAttributeName({ endpointId, clusterId, attributeId }: TypeFromSchema<typeof TlvAttributePath>) {
```

Comment:
OK now there's really getting to be a lot of duplicate code.  Ideally maybe move to a generic "ElementPath" type where the path becomes `{ elementType, endpointId, clusterId, attributeId }`?

If that's too big of a change, maybe at least pull endpoint/cluster lookup into a this.resolveCluster(endpointId, clusterId) function.

Reply:
I was able to generalize this ... see next commit